Add field binning to the binning plugin (#5319)
ComputationalRadiationPhysics/picongpu: docs/source/usage/plugins/binningPlugin.rst

The functor needs to follow the signature shown below. This provides the user access to the particle or field object and with information about the :ref:`domain <usage/plugins/binningPlugin:Domain Info>`.

The return type is defined by the user.

For particles:

.. code-block:: c++

	auto myParticleFunctor = [] ALPAKA_FN_ACC(auto const& worker, auto const& domainInfo, auto const& particle) -> returnType
	{
		// fn body
		return myParameter;
	};

For particles with two extra data paramters:

.. code-block:: c++

	auto myParticleWithExtraDataFunctor = [] ALPAKA_FN_ACC(auto const& worker, auto const& domainInfo, auto const& particle, auto const& extraData1, auto const& extraData2) -> returnType
	{
		// fn body
		return myParameter;
	};

For fields:

.. code-block:: c++

	auto myFieldFunctor = [] ALPAKA_FN_ACC(auto const& worker, auto const& domainInfo, auto const& field) -> returnType
	{
		// fn body
		return myParameter;
	};

For fields with extra data:

.. code-block:: c++

	auto myFieldWithExtraDataFunctor = [] ALPAKA_FN_ACC(auto const& worker, auto const& domainInfo, auto const& field1, auto const& field2, auto const& extraData) -> returnType
	{
		// fn body
		return myParameter;

> 	};

Reviewer: This collection of examples is not quite clear to me. Would you mind providing more guidance through those examples? What's specific to particles/fields/... and what's general? What would be a context to find those signatures in?

Author: Added some descriptive sentences. Let me know if they help.

.. note::

	Fields and extra data may be tuples with multiple elements which are unpacked and passed to the user-defined functors. Therefore, it is the responsibility of the user to ensure that their functors have an appropriate number of arguments to match the provided tuples.

Domain Info
"""""""""""
Enables the user to find the location of the particle or field in the simulation domain. Contains

For particle binning, the ``DomainInfo`` class contains:

.. doxygenclass:: picongpu::plugins::binning::DomainInfoBase
	:members:

For field binning, the ``DomainInfo`` class additionally contains:

.. doxygenclass:: picongpu::plugins::binning::DomainInfo< BinningType::Field >

> 	:members:

Reviewer: This does not render into something useful.

> The global and local offsets can be understood by looking at the `PIConGPU domain definitions <https://github.com/ComputationalRadiationPhysics/picongpu/wiki/PIConGPU-domain-definitions>`_.

Reviewer: We should definitely bring this information into read the docs! (Not actionable in this PR.)

Author: Agreed! Will add info in #5333

For particle binning, the particle position is obtained at cell precision by default. To get sub-cell precision or SI units, use optional template parameters with ``getParticlePosition<DomainOrigin, PositionPrecision, PositionUnits>``.

> For field binning, the field ``DomainInfo`` additionally holds the localCellIndex in the supercell and has a method to ``getCellIndex<DomainOrigin, PositionUnits>`` to get the current cell index being binned relative to an origin (global, total, local). To get the exact position of the fields inside the cell, relative to the cell index, use the FieldPosition trait.

Reviewer: Should the `FieldPosition trait` have a link?

Functor Description
-------------------
The basic building block for the binning plugin is the Functor Description object, and it is used to describe both axes and the deposited quantity.
It describes the properties which we find interesting and how we can calculate/get this property from the particle or field.
A functor description is created using createFunctorDescription.

.. doxygenfunction:: picongpu::plugins::binning::createFunctorDescription

Functor
^^^^^^^

> The functor is run on the GPU and is how the user can execute code on device while binning. This may be used to calculate the deposited quantity or to calculate the axis values.

Reviewer: ```suggestion
The functor is run on the GPU and constitutes the means by which the user can execute code on device while binning. This may be used to calculate the deposited quantity or to calculate the axis values.
```

Author: Fixed

The functor needs to follow the signature shown below. This provides the user access to the particle or field object and with information about the :ref:`domain <usage/plugins/binningPlugin:Domain Info>`.

The return type is defined by the user.

> The first two arguments are always the worker and the domainInfo. For particle binning this is followed bt the particle object, and for field binning this is followed by the field object (if any). If extra data is passed, this is also passed as arguments.

Reviewer: ```suggestion
The first two arguments are always the worker and the domainInfo. For particle binning this is followed by the particle object, and for field binning this is followed by the field object (if any). If extra data is passed, this is also passed as arguments.
```

Author: fixed

It is the responsibility of the user to ensure that their functors have an appropriate number and types of arguments to match the provided tuples. A few examples are shown below.

For particles:

.. code-block:: c++

	auto myParticleFunctor = [] ALPAKA_FN_ACC(auto const& worker, auto const& domainInfo, auto const& particle) -> returnType
	{
		// fn body
		return myParameter;
	};

For particles with two extra data paramters:

.. code-block:: c++

	auto myParticleWithExtraDataFunctor = [] ALPAKA_FN_ACC(auto const& worker, auto const& domainInfo, auto const& particle, auto const& extraData1, auto const& extraData2) -> returnType
	{
		// fn body
		return myParameter;
	};

For one field:

.. code-block:: c++

	auto myFieldFunctor = [] ALPAKA_FN_ACC(auto const& worker, auto const& domainInfo, auto const& field) -> returnType
	{
		// fn body
		return myParameter;
	};

For two fields with one extra data parameter:

.. code-block:: c++

	auto myFieldWithExtraDataFunctor = [] ALPAKA_FN_ACC(auto const& worker, auto const& domainInfo, auto const& field1, auto const& field2, auto const& extraData) -> returnType
	{
		// fn body
		return myParameter;

> 	};

Reviewer: That is much better readable. Could we also do `literalincludes` here? (Meaning: Do we have sufficiently diverse examples in order to show this variety?)

Author: I don't think we have sufficiently diverse examples, especially as I want to keep the ``literalinclude`` labels in the compile2 file to avoid clutter in files more often seen by users. Furthermore, I think these short code snippets are quite instructive.

Reviewer: I disagree that this form of minimal clutter is something to worry about but it's fine. There is some value in having these completely stripped down examples.

.. note::

	Fields and extra data may be tuples with multiple elements which are unpacked and passed to the user-defined functors. Therefore, it is the responsibility of the user to ensure that their functors have an appropriate number of arguments to match the provided tuples.

Domain Info
"""""""""""
Enables the user to find the location of the particle or field in the simulation domain. Contains

For particle binning, the ``DomainInfo`` class contains:

.. doxygenclass:: picongpu::plugins::binning::DomainInfoBase
	:members:

For field binning, the ``DomainInfo`` class additionally contains:

.. doxygenfunction:: picongpu::plugins::binning::DomainInfo<BinningType::Field>::getCellIndex

> 	:members:

Reviewer: This vanished completely during rendering.

Reviewer: No we get
```
doxygenfunction: Cannot find function “picongpu::plugins::binning::DomainInfo<BinningType::Field>::getCellIndex” in doxygen xml output for project “PIConGPU” from directory: ../xml
```

I suggest you build the docs locally and debug this seriously.

Author: And i have made it vanish from the source as well. I gave up on fighting with doxygen. The function is mentioned in the text below anyways.

.. code-block:: c++

	std::array<double, numUnits> momentumDimension{};
	momentumDimension[SIBaseUnits::length] = 1.0;
	momentumDimension[SIBaseUnits::mass] = 1.0;

> 	momentumDimension[SIBaseUnits::time] = -1.0;

Reviewer: Please use an excerpt from tested files instead of hard-coding code snippets.

Author: done

	:members:

Range
"""""

.. doxygenclass:: picongpu::plugins::binning::axis::Range
	:members:

.. note::

    Axes are passed to addParticleBinner or addFieldBinner grouped in a tuple. This is just a collection of axis objects and is of arbitrary size.
    Users can make a tuple for axes by using the ``createTuple()`` function and passing in the axis objects as arguments.

Species
-------
PIConGPU species which should be used in particle binning.
Species can be instances of a species type or a particle species name as a PMACC_CSTRING. For example,

.. code-block:: c++

	auto electronsObj = PMACC_CSTRING("e"){};

Optionally, users can specify a filter to be used with the species. This is a predicate functor, i.e. it is a functor with a signature as described above and which returns a boolean. If the filter returns true it means the particle is included in the binning.
They can then create a FilteredSpecies object which contains the species and the filter.

.. code-block:: c++

	auto myFilter = [] ALPAKA_FN_ACC(auto const& worker, auto const& domainInfo, auto const& particle) -> bool
	{
		bool binningEnabled = true;
		// fn body
		return binningEnabled;
	};

> 	auto fitleredElectrons = FilteredSpecies{electronsObj, myFilter};

Reviewer: Same here. Refer to tested code.

Author: done

.. note::

			Species are passed to addParticleBinner in the form of a tuple. This is just a collection of Species and FilteredSpecies objects (the tuple can be a mixture of both) and is of arbitrary size.
			Users can make a species tuple by using the ``createSpeciesTuple()`` function and passing in the objects as arguments.

Fields
------
PIConGPU fields which should be used in field binning.
Fields can be instances of a field type. For example,

.. code-block:: c++

	DataConnector& dc = Environment<>::get().DataConnector();
	auto fieldsTuple = createTuple(
	dc.get<FieldB>(FieldB::getName())->getDeviceDataBox(),

> 	dc.get<FieldTmp>(FieldTmp::getUniqueId(0))->getDeviceDataBox());

Reviewer: Everywhere... Won't comment this anymore.

Author: Done

Fields are passed to addFieldBinner in the form of a tuple. This is just a collection of field objects and is of arbitrary size.
Users can make a fields tuple by using the ``createTuple()`` function and passing in the objects as arguments.

.. note::

	It is possible to have an empty tuple for fields when doing field binning, in which case the functor will be called with no fields. This may be useful if you are passing in extra data and want field traversal over it.

Deposited Quantity
------------------
Quantity to be deposited is simply a :ref:`functor description <usage/plugins/binningPlugin:Functor Description>`.
The signature of quantity functors is

.. code-block:: c++

	auto myQuantityFunctor = [] ALPAKA_FN_ACC(auto const& worker, auto const& domainInfo, ...) -> returnType

Extra Data
----------
Users can pass extra data to the functors if additional information is required by the kernels to do binning.

> 	This extra data is created using the `createTuple()` function and then passed into the binning creator.

Reviewer: ```suggestion
This extra data is created using the `createTuple()` function and then passed into the binning creator.
```

Author: done

Extra data is passed to ``addParticleBinner`` or ``addFieldBinner`` functions in the form of a tuple. This tuple is just a collection of objects and is of arbitrary size.

> Users can make a tuple of the extra data by using the ``createTuple()`` function and passing in the objects as arguments.

Reviewer: There's some duplication here.

Author: Fixed

writeOpenPMDFunctor
-------------------
Users can also write out custom output to file, for example other quantities related to the simulation which the user is interested in.
This is a lambda with the following signature, and is set using the ``setWriteOpenPMDFunctor`` method.

.. code-block:: c++

	[=](::openPMD::Series& series, ::openPMD::Iteration& iteration, ::openPMD::Mesh& mesh) -> void

.. note::

> 	Make sure to capture by copy only, as the objects defined in the param file are not kept alive

Reviewer: If you want to transfer ownership, you should probably use an explicit move passing them to the functor. That way the user will also get a "clear" (in whatever twisted way you want to apply this attribute to C++ error messages) error message that they can't do that.

Author: Hmm. I'm not sure whats the best way to explain lifetimes to users here. I don't want to recommend moving here since it is not clear if the user is ready to transfer ownership immidiately when this functor is encountered.

Reviewer: Sorry, I think I misunderstood. You are referring to the `=` sign in the lambda capture, right? I was thinking about the signature of the lambda.

        *Calibration and cross-laboratory implementation of scintillating screens for electron bunch charge determination*,
        Review of Scientific Instruments (2018),
        https://doi.org/10.1063/1.5041755

More examples of binning plugin usage
=====================================

> Further examples of the binning plugin can be found in the ``binningSetup.param`` associated with the atomic physics example and the ``compile2`` test.

Reviewer: Do you think this would benefit from having links?

Author: Changed to full relative paths based on what is currently being done in the rest of the docs.
